cli.js

```diff
@@ -0,0 +1,180 @@
+    }
+    return slurpStream(istream);
+}; // }}}1
+
+let getPort = (value, dummyPrevious) => { // {{{1
+    // Get a valid port number from the command line
+    let ret = parseInt(value, 10);
+    if (ret < 1 || ret > 65535) {
+        throw new Error("invalid port number");
+    }
+    return ret;
+};
+// }}}1
+
+///////// MAIN ///////// {{{1
+
+const chef = require("cyberchef");
+const program = require("commander");
+
+program
+    .version(require('./package.json').version)
+    .description('Bake data from files and/or TCP clients '
+        + 'using a CyberChef recipe.')
+    .usage('[options] [file [file ...]]')
+    .requiredOption('-r, --recipe-file <file>',
+        'recipe JSON file')
+    .option('-l, --listen [port]',
+        'listen on TCP port for data (random if not given)', getPort, false)
+    .option('-o, --output <file-or-dir>',
+        'where to write result (file input only; default:stdout)');
+
+try {
+    program.exitOverride().parse(process.argv);
+}
+catch (e) {
+    if (e.code != 'commander.helpDisplayed') {
+        console.error("Run with '--help' for usage");
+    }
+    process.exit(1);
+}
+
+// If we get no inputs and we aren't running a server,
+// make stdin our single input
+let inputs = program.args;
+if (inputs.length == 0 && !program.listen) {
+    inputs = [ '-' ];
+}
+
+// Likewise stdout for our output
+let ostream;
+let path;
+let outputIsDir = false;
+if (program.output === undefined && !program.listen) {
+    ostream = process.stdout;
+}
+else if (inputs.length > 0) {
+    // See if our output is a directory
+    let st;
+    try {
+        st = fs.statSync(program.output);
+        outputIsDir = st.isDirectory();
+    }
+    catch(err) {
+        // We're fine if the output doesn't exist yet
+        if (err.code != 'ENOENT') throw err;
+    }
+    if (!outputIsDir) {
+        ostream = fs.createWriteStream(program.output);
+    }
+}
+if (outputIsDir) path = require("path");
+
+let recipe;
+slurp(program.recipeFile).then((data) => {
+    recipe = JSON.parse(data);
+})
+.catch((err) => {
+    console.error(`Error parsing recipe: ${err.message}`);
+    process.exit(2);
+})
+.then(() => {
+    // First, deal with any files we want to read
+    for(let i of inputs) {
+        slurp(i).then((data) => {
+            let output = chef.bake(data, recipe);
+            if (outputIsDir) {
+                let outFileName = path.basename(i);
+                if (outFileName == '-') outFileName = 'from-stdin';
+                ostream = fs.createWriteStream(
+                    path.join(program.output, outFileName));
+            }
+            ostream.write(output.presentAs("string", true));
+            if (outputIsDir) ostream.end();
+        },
+        (err) => {
+            console.error(err.message);
+            process.exitCode = 2;
+        })
+        .catch((err) => {
+            console.error(err.message);
+            process.exitCode = 2;
+        });
```

Comment:
One quirk of Node I haven't yet been able to work out is why I need the failure handler on this `then()` call in addition to the `catch()` call.

Reply:
You don't need the failure handler on it, you can either use that or [`.catch()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch).